Source/DafnyLanguageServer/Workspace/TextDocumentLoader.cs

```diff
@@ -184,126 +206,103 @@ bool loadCanceled
-          HandleStatusUpdate(implementationTask, boogieStatus).ToObservable());
-      });
-      var result = implementationsUpdates.Merge().Replay();
-      result.Connect();
-      return result;
+      try {
+        var observable = implementationTask.Run();
+        cancellationToken.Register(() => {
+          try {
+            implementationTask.Cancel();
+          } catch (InvalidOperationException e) {
+            if (!e.Message.Contains("no ongoing run")) {
+              throw;
+            }
+          }
+        });
+        return GetVerifiedDafnyDocuments(dafnyDocument, implementationTask, observable);
+      } catch (InvalidOperationException e) {
+        if (e.Message.Contains("already completed") || e.Message.Contains("ongoing run")) {
+          return Observable.Empty<DafnyDocument>();
+        }
 
+        throw;
+      }
+    }
 
-      async Task<DafnyDocument> HandleStatusUpdate(IImplementationTask implementationTask, VerificationStatus boogieStatus) {
-        var id = GetImplementationId(implementationTask.Implementation);
-        var status = await StatusFromImplementationTaskAsync(implementationTask);
-        var lspRange = implementationTask.Implementation.tok.GetLspRange();
-        if (boogieStatus is VerificationStatus.Completed) {
-          var verificationResult = await implementationTask.ActualTask;
-          foreach (var counterExample in verificationResult.Errors) {
-            counterExamples.Push(counterExample);
-          }
+    private IObservable<DafnyDocument> GetVerifiedDafnyDocuments(DafnyDocument document, IImplementationTask implementationTask,
+      IObservable<IVerificationStatus> statusUpdates) {
 
-          var itDiagnostics = GetDiagnosticsFromResult(document, verificationResult);
-          var view = new ImplementationView(lspRange, status, itDiagnostics);
-          implementationViews.AddOrUpdate(id, view, (_, _) => view);
-          if (VerifierOptions.GutterStatus) {
-            progressReporter.ReportEndVerifyImplementation(implementationTask.Implementation, verificationResult);
-          }
-        } else {
-          implementationViews.AddOrUpdate(id,
-            _ => new ImplementationView(lspRange, status, Array.Empty<Diagnostic>()),
-            (_, previousView) => previousView with { Status = status });
+      var result = statusUpdates.Select(boogieStatus => HandleStatusUpdate(document, implementationTask, boogieStatus));
+
+      var initial = document with {
+        ImplementationViewsView = document.ImplementationViews!.ToImmutableDictionary(),
+      };
+      return Observable.Return(initial).Concat(result);
+    }
+
+    DafnyDocument HandleStatusUpdate(DafnyDocument document, IImplementationTask implementationTask, IVerificationStatus boogieStatus) {
+      var id = GetImplementationId(implementationTask.Implementation);
+      var status = StatusFromBoogieStatus(boogieStatus);
+      var lspRange = implementationTask.Implementation.tok.GetLspRange();
```

Comment:
```suggestion
      var implementationRange = implementationTask.Implementation.tok.GetLspRange();
```

In LSP, it's mostly clear that ranges are always LSP ranges, so you can name them more precisely according to their function.